### ISSUE TITLE
lstsq/svd always uses the full quadratic memory: should sometimes use `cuSolverDnSgesvdp` or similar

### ISSUE THREAD

**Issue by rgreenblatt** [google/jax#9793]
Using lstsq on moderately large inputs runs out of memory allocating a buffer of size `m^2`
For instance:
```python
import numpy as np
import jax
import jax.numpy as jnp

x = np.arange(100_000_000, dtype=np.float32)[:, None]
y = x * 3
print(x.shape)

print(np.linalg.lstsq(x, y, rcond=None))  # totally fine
print(jax.jit(lambda x, y: jnp.linalg.lstsq(x, y, rcond=None), backend="cpu")(x, y))  # wrong? but doesn't OOM
print(jnp.linalg.lstsq(x, y, rcond=None)[0])  # OOM
```
This comes down to the svd implementation used for the gpu backend (always `cuSolverDnSgesvd`):
```python
print(jnp.linalg.svd(x, full_matrices=False))  # OOM
```

When `full_matrices=False` and `x` is far from square svd can use much less memory via `cuSolverDnSgesvdp` (or maybe some other cuSolver function).
It might also be nice to support low rank svd via `cuSolverDnSgesvdr` like `torch.svd_lowrank`.

Additionally, it looks like the cpu implementation is broken? It should get a coefficient of 3 in the example above, but it actually returns 0 with a large residual.

**Comment by hawkinsp** [comment]
The `full_matrices` issue on GPU is already fixed (https://github.com/google/jax/pull/9760), although the fix isn't yet released as a wheel, so to get it you'd have to build from source.

I'll look at the CPU issue.

**Comment by hawkinsp** [comment]
I think the CPU issue stems from low precision for matrix multiplications with very large contracting dimensions on CPU. The workaround for now would be to enable x64 mode and use 64-bit precision.

```
In [45]: z = jnp.ones((100000000, 1), jnp.float32)

In [46]: z.T  @ z
Out[46]: DeviceArray([[16777216.]], dtype=float32)

In [47]: z = jnp.ones((100000000, 1), jnp.float64)

In [48]: z.T  @ z
Out[48]: DeviceArray([[1.e+08]], dtype=float64)
```

**Comment by rgreenblatt** [comment]
> The `full_matrices` issue on GPU is already fixed (#9760), although the fix isn't yet released as a wheel, so to get it you'd have to build from source.
> 
> I'll look at the CPU issue.

Nice. I wonder how I missed this issue when searching.
